table/tables/partition.go

```diff
@@ -263,6 +267,49 @@ type PartitionExpr struct {
 	*ForListPruning
 }
 
+// GetPartColumnsForKeyPartition is used to get partition columns for key partition table
+func (pe *PartitionExpr) GetPartColumnsForKeyPartition(columns []*expression.Column) ([]*expression.Column, []int) {
+	schema := expression.NewSchema(columns...)
+	partCols := make([]*expression.Column, len(pe.ColumnOffset))
+	colLen := make([]int, 0, len(pe.ColumnOffset))
+	for i, offset := range pe.ColumnOffset {
+		partCols[i] = schema.Columns[offset]
+		partCols[i].Index = i
+		colLen = append(colLen, partCols[i].RetType.GetFlen())
+	}
+	return partCols, colLen
+}
+
+// LocateKeyPartitionWithSPC is used to locate the destination partition for key
+// partition table has single partition column(SPC). It's called in FastPlan process.
+func (pe *PartitionExpr) LocateKeyPartitionWithSPC(pi *model.PartitionInfo,
+	r []types.Datum) (int, error) {
+	col := &expression.Column{}
+	*col = *pe.KeyPartCols[0]
+	col.Index = 0
+	return pe.LocateKeyPartition(pi, []*expression.Column{col}, r)
+}
+
+// LocateKeyPartition is the common interface used to locate the destination partition
+func (pe *PartitionExpr) LocateKeyPartition(pi *model.PartitionInfo,
+	cols []*expression.Column, r []types.Datum) (int, error) {
+	nr1 := uint64(1)
+	nr2 := uint64(4)
+	for _, col := range cols {
+		if r[col.Index].Kind() == types.KindNull {
+			nr1 ^= (nr1 << 1) | 1
+			continue
+		}
+		data, err := r[col.Index].ToBytes()
```

Comment:
Have you tested with any unique/primary key that is [var]char of a non-binary collation, like utf8mb4_general_ci?
I would guess that with just getting the bytes directly from the column would actually allow duplicate keys, since collations compare data differently, like 'a' == 'A', but `ToBytes()` + hash will give different values. Maybe use Collator.Key() function?

Reply:
Do you say using crc32 to test? or using current hash algorithm?  if use crc32, it will give different values for 'a' and 'A' even using utf8mb4_general_ci, but we can't insert a duplicated key I think, because the unque check is not guaranted by the hash value.
So I recommend that using current hash algoritm, which will ensure "A" and "a" placed in the same partition.
I'll add a case for using utf8mb4_general_ci character set.

Reply:
I have tested the performance of the two hash algorithm, as follows. I exectuted every hash algorithm for 10 millions times, it's the average time every millions executing.
x86:
LocateKeyPartition-selfhash: 82 ms
LocateKeyPartition-crc32: 123 ms
arm:
LocateKeyPartition-selfhash: 106
LocateKeyPartition-crc32: 118

The crc32 hash algorithm is slower than selfhash.

Reply:
[test.txt](https://github.com/pingcap/tidb/files/10790636/test.txt)

Reply:
I added a comment about the added test case, where I show that current code does not work with case insensitive collation.

I don't think the time difference between standard crc32 and the simple hash function is too big, compared to the benefits of better distribution. I would also assume that that just adding the column loop, check for Null, check for collation and convert ToBytes()/Key() will make the difference even less.

By the way, you could add a benchmark test like BenchmarkRangeColumnsPruner100 to make sure that the performance is affected drastically by future changes.

Reply:
ok, I agree with you, Thank you very much